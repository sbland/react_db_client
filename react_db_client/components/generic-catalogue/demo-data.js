/* eslint-disable react/prop-types */
import React, { useEffect, useMemo } from 'react';
import { comparisons, filterTypes } from '@samnbuk/react_db_client.constants.client-types';

export const demoHeadingsDataSimple = [
  {
    uid: 'uid',
    label: 'UID',
    type: 'button',
    action: (x) => alert(x),
  },
  {
    uid: 'name',
    label: 'Name',
    type: filterTypes.text,
  },
  {
    uid: 'count',
    label: 'Count',
    type: filterTypes.number,
    defaultValue: 7,
    step: 1,
  },
  {
    uid: 'eval',
    label: 'Eval',
    evaluateType: 'number',
    type: filterTypes.number,
    expression: '$count + 1',
    expressionReversed: '$count=$_ - 1',
  },
];

export const demoResults = {
  a: {
    uid: 'a',
    natid: '100a',
    name: 'Foo',
    count: 3,
    multiplier: 2,
    toggle: true,
    readOnly: "Can't touch me",
    select: 'a',
    hiddenDemoNumber: 3,
    hiddenDemo: 'Hide me',
  },
  ab: {
    // Note string "4" here to check we can pass string numbers without issues
    uid: 'ab',
    natid: '10a',
    name: 'Foobar',
    count: '99',
    def: 3,
    hiddenDemoNumber: 3,
    hiddenDemo: 'Hide me',
  },
  b: {
    uid: 'b',
    natid: '50a',
    name: 'Bar',
    description: 'A really really really long description that needs to fit into a little box!',
    hiddenDemoNumber: 3,
    hiddenDemo: 'Hide me',
  },
  c: {
    uid: 'c',
    name: 'C',
    count: 3,
    multiplier: 3,
    expression: 9,
    hiddenDemoNumber: 3,
    hiddenDemo: 'Hide me',
  },
  d: {
    uid: 'd',
    name: '',
    hiddenDemoNumber: 3,
    hiddenDemo: 'Hide me',
  },
};

export const demoHeadingsData = [
  {
    uid: 'uid',
    label: 'UID',
    type: 'button',
    action: (x) => alert(x),
    styleRule: '$count<5',
    unique: true,
    required: true,
    group: 0,
  },
  {
    uid: 'name',
    label: 'Name',
    type: 'text',
    unique: true,
    required: true,
    group: 0,
  },
  {
    uid: 'description',
    label: 'description',
    type: 'textLong',
    group: 1,
  },
  {
    uid: 'count',
    label: 'Count',
    type: 'number',
    max: 10,
    min: 2,
    def: 8,
    step: 1,
    showTotals: true,
    group: 3,
  },
  {
    uid: 'def',
    label: 'Def',
    type: 'number',
    max: 10,
    min: 1,
    defaultValue: 7,
    showTotals: true,
    group: 3,
  },
  {
    uid: 'multiplier',
    label: 'Multiplier',
    type: 'number',
    max: 10,
    min: 1,
    def: '$def',
    group: 4,
  },
  {
    uid: 'readOnly',
    label: 'Read Only',
    type: 'text',
    readOnly: true,
    group: 4,
  },
  {
    uid: 'hiddenDemo',
    label: 'Hidden',
    type: 'text',
    hidden: true,
    group: 5,
  },
  {
    uid: 'hiddenDemoNumber',
    label: 'Hidden Number',
    type: 'number',
    hidden: true,
    group: 5,
  },
  {
    uid: 'toggle',
    label: 'Toggle',
    type: 'toggle',
    group: 5,
  },
  {
    uid: 'toggleR',
    label: 'ToggleR',
    type: 'toggle',
    readOnly: true,
    group: 5,
  },
  {
    uid: 'expression',
    label: 'Expression',
    type: 'number',
    evaluateType: 'number',
    expression: '$count * $multiplier',
    expressionReversed: '$count=$_ / $multiplier',
    reverseExpression: '$count * $multiplier',
    showTotals: true,
    group: 5,
  },
  {
    uid: 'validated',
    label: 'Is Validated',
    evaluateType: 'string',
    type: 'string',
    operator: comparisons.equals,
    target: 'a',
    field: 'select',
    invert: true,
    group: 5,
  },
  {
    uid: 'invalid',
    label: 'Invalid',
    evaluateType: 'number',
    type: 'number',
    expression: '$nothing',
    group: 5,
  },
  {
    uid: 'invalidb',
    label: 'InvalidB',
    type: 'number',
    evaluateType: 'number',
    expression: '$invalid * 2',
    group: 5,
  },
  {
    uid: 'select',
    label: 'Select',
    type: 'select',
    options: [
      { uid: 'a', label: 'A' },
      { uid: 'b', label: 'B' },
      { uid: 'c', label: 'C' },
    ],
    group: 5,
  },
  {
    uid: 'internalRoom',
    label: 'Internal Room',
    type: 'entity',
    collection: 'rooms',
    group: 5,
  },
  {
    uid: 'longHeading',
    label: 'Long Heading which is really really long',
    type: 'entity',
    collection: 'rooms',
    group: 5,
  },
  {
    uid: 'customfield',
    label: 'Custom',
    type: 'custom',
    isCustom: true,
    showTotals: true,
    group: 5,
  },
  {
    uid: 'customfieldeval',
    label: 'Custom Eval',
    type: 'customeval',
    isCustom: true,
    showTotals: true,

    expression: '$customfield + 1',
    expressionReversed: '$customfield=$_ - 1',
    evaluateType: 'number',
    group: 5,
  },
];

export const CustomField = ({ acceptValue, cellData, editMode, focused }) => {
  const actionFunc = useMemo(() => {
    return () => acceptValue(Math.random());
  }, [acceptValue]);

  useEffect(() => {
    if (editMode && focused) {
      actionFunc();
    }
  }, [editMode, focused, actionFunc]);
  return (
    <div className="dataTableCellData dataTableCellData-button">
      <button type="button" onClick={actionFunc}>
        {cellData}
      </button>
    </div>
  );
};

// eslint-disable-next-line no-unused-vars
export const customFilter = (value, expression, targetValue, item) => {
  if (Math.abs(value - targetValue) < 0.3) return true;
  return false;
};

// =================================
// import { filterTypes } from '@samnbuk/react_db_client.constants.client-types';

// export const demoHeadingsData = [
//   {
//     uid: 'choice',
//     label: 'Choice',
//     type: filterTypes.select,
//     options: [
//       { uid: 'choicea', label: 'Choice A' },
//       { uid: 'choiceb', label: 'Choice B' },
//     ],
//     required: true,
//     group: 0,
//   },
//   {
//     uid: 'cost',
//     label: 'Cost',
//     type: filterTypes.number,
//     required: true,
//     group: 1,
//   },
//   {
//     uid: 'code',
//     label: 'Code',
//     type: filterTypes.text,
//     required: true,
//     group: 2,
//   },
//   {
//     uid: 'description',
//     label: 'Description',
//     type: filterTypes.textLong,
//     group: 2,
//   },
//   {
//     uid: 'images',
//     label: 'Images',
//     multiple: true,
//     type: filterTypes.image,
//     group: 5,
//   },
//   {
//     uid: 'val',
//     label: 'Value',
//     type: filterTypes.number,
//     group: 5,
//   },
//   {
//     uid: 'documentation',
//     label: 'Documentation',
//     type: filterTypes.file,
//     fileType: 'document',
//     multiple: true,
//     group: 6,
//   },
//   {
//     uid: 'thumbnail',
//     label: 'Thumbnail',
//     type: filterTypes.file,
//     fileType: 'image',
//     group: 7,
//   },
// ];

// export const demoHeadingsDataSimple = demoHeadingsData.slice(0, 4);

// export const demoResults = {
//   demoid: {
//     uid: 'demoid',
//     choice: 'choicea',
//     cost: 10,
//     code: 'Code',
//     description: 'A long description',
//     images: [],
//     val: 9,
//     documentation: 'doc.pdf',
//     thumbnail: 'thumbnail.jpg',
//   },
// };
