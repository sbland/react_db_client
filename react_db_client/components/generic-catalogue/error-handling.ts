export class GenericCatalogueError extends Error {
  name: 'GenericCatalogueError';
  error?: Error;
  constructor(msg: string, error?: Error) {
    super(msg);
    this.error = error;
    // Set the prototype explicitly.
    Object.setPrototypeOf(this, GenericCatalogueError.prototype);
  }
}
