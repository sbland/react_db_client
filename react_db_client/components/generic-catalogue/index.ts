export { GenericCatalogue } from './generic-catalogue';
export type { IGenericCatalogueProps, TResultHeading } from './generic-catalogue';
export { GenericCatalogueError } from './error-handling';
