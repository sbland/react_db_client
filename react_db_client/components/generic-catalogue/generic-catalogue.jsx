import React, { useState, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';

import { SearchAndSelect } from '@samnbuk/react_db_client.components.search-and-select';
import { AreYouSureBtn } from '@samnbuk/react_db_client.components.are-you-sure-btn';
import { ItemEditor as _ItemEditor } from '@samnbuk/react_db_client.components.item-editor';
import { Emoji } from '@samnbuk/react_db_client.components.emoji';


export const GenericCatalogue = ({
  itemName,
  collection,
  additionalFilters,
  resultsHeadings,
  editorHeadings,
  additionalSaveData,
  availableFilters,
  ItemEditor,
  errorCallback,
  apiGetDocuments,
  apiDeleteDocument,
  PopupPanel,
}) => {
  const [showEditor, setShowEditor] = useState(false);
  const [selectedUid, setSelectedUid] = useState(null);
  const [reloadDatakey, setReloadDataKey] = useState(0);
  const [handleDelete, setHandleDelete] = useState(false);

  // const { call: copySet } = useAsyncRequest({
  //   args: [],
  //   callFn: apiCopyDocument,
  //   callOnInit: false,
  //   callback: () => {
  //     setReloadDataKey((prev) => prev + 1);
  //     alert('Successfully Copied Set');
  //   },
  // });

  useEffect(() => {
    if (handleDelete) {
      setHandleDelete(false);
      apiDeleteDocument(collection, selectedUid)
        .then(() => {
          setReloadDataKey((prev) => prev + 1);
        })
        .catch(() => {
          errorCallback(`Error deleting ${itemName}`, 'error');
        });
    }
  }, [handleDelete, selectedUid, errorCallback, collection, itemName]);

  // TODO: Implement duplicate
  // const handleDuplicate = (uid) => {
  //   const fromCollection = COLLECTION;
  //   const fromUid = uid;
  //   const toCollection = COLLECTION;
  //   const toUid = generateUid(COLLECTION);
  //   const additionalData = { label: toUid };
  //   copySet([fromCollection, fromUid, toCollection, toUid, additionalData]);
  // };

  const searchFn = useCallback(
    async (filters, sortBy, searchString) => {
      const filtersModified = [...additionalFilters, ...filters];
      const sortByA = sortBy || 'label';
      const docs = await apiGetDocuments(
        collection,
        filtersModified,
        // TODO: Filter preview only headings out
        resultsHeadings.map((item) => item.uid),
        sortByA,
        searchString,
        false,
        'all'
      );
      return docs;
    },
    [collection, additionalFilters, resultsHeadings]
  );

  return (
    <>
      {showEditor && (
        <ItemEditor
          inputUid={selectedUid}
          isNew={!selectedUid && true}
          additionalData={additionalSaveData}
          params={editorHeadings}
          collection={collection}
          onSubmitCallback={(uid) => {
            notificationDispatch(openMessage(`${itemName} saved`));
            setSelectedUid(uid);
            setShowEditor(false);
            setReloadDataKey(reloadDatakey + 1);
          }}
        />
      )}
      <div className="GenericCatalogueFunc_Wrap sectionWrapper">
        <section>
          <button
            type="button"
            className="button-two"
            onClick={() => {
              setSelectedUid(null);
              setShowEditor(true);
            }}
          >
            Create New {itemName}
          </button>
        </section>
        <section style={{ width: '100%' }}>
          <p>
            <b>Adjust the filters below to search for {itemName}s</b>
          </p>
          <SearchAndSelect
            key={reloadDatakey}
            searchFunction={searchFn}
            initialFilters={additionalFilters}
            availableFilters={availableFilters}
            handleSelect={setSelectedUid}
            // TODO: Filter preview only headings
            headings={resultsHeadings}
            previewHeadings={resultsHeadings}
            autoWidth={false}
            autoUpdate
            showRefreshBtn
            showSearchField
            loadOnInit={false}
            noEmptySearch
            allowSelectionPreview
          />
          <div>
            <button
              type="button"
              onClick={() => setShowEditor(true)}
              disabled={selectedUid === null}
            >
              Edit Selected {itemName}
            </button>
            <AreYouSureBtn
              onConfirmed={() => {
                setHandleDelete(true);
              }}
              message={`Delete ${itemName}`}
              disabled={selectedUid === null}
              btnText={<Emoji emoj="🗑️" label="Delete" />}
              PopupPanel={PopupPanel}
            />
          </div>
        </section>
      </div>
    </>
  );
};

GenericCatalogue.propTypes = {
  itemName: PropTypes.string.isRequired,
  collection: PropTypes.string.isRequired,
  additionalFilters: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      field: PropTypes.string.isRequired,
      operator: PropTypes.string,
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    })
  ),
  resultsHeadings: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  editorHeadings: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  additionalSaveData: PropTypes.shape({}),
  availableFilters: PropTypes.objectOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
      required: PropTypes.bool,
      group: PropTypes.number,
    })
  ).isRequired,
  ItemEditor: PropTypes.elementType,
  apiGetDocuments: PropTypes.func.isRequired,
  apiDeleteDocuments: PropTypes.func.isRequired,
  PopupPanel: PropTypes.elementType,
};

GenericCatalogue.defaultProps = {
  additionalFilters: [],
  additionalSaveData: {},
  ItemEditor: _ItemEditor,
  PopupPanel: () => {},
};
