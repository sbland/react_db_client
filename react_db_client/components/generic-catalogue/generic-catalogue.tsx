import React, { useState, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';

import { TComponentMap } from '@form-extendable/lib';
import {
  CustomParser,
  IHeading,
  ISearchAndSelectProps,
  SearchAndSelect,
} from '@react_db_client/components.search-and-select';
import { AreYouSureBtn } from '@react_db_client/components.are-you-sure-btn';
import {
  FilterObjectClass,
  FilterOption,
  IDocument,
  TAsyncCopyDocument,
  TAsyncDeleteDocument,
  TAsyncGetDocument,
  TAsyncGetDocuments,
  TAsyncPostDocument,
  TAsyncPutDocument,
  Uid,
} from '@react_db_client/constants.client-types';
import { ItemEditor as ItemEditorDefault } from '@react_db_client/components.item-editor';
import { useAsyncRequest } from '@react_db_client/async-hooks.use-async-request';
import { Emoji } from '@react_db_client/components.emoji';
import { generateUid } from '@react_db_client/helpers.generate-uid';

// Link these to actual types
export interface IItemEditorProps {}
export interface IPopupPanelProps {}

export interface IGenericCatalogueProps<ResultType extends IDocument> {
  id: Uid;
  itemName: string;
  collection: string;
  additionalFilters?: FilterObjectClass[];
  customSort?: (a: ResultType, b: ResultType) => -1 | 0 | 1;
  resultsHeadings: IHeading[];
  editorHeadings: IHeading[];
  additionalSaveData?: object;
  availableFilters: { [key: string]: FilterOption<any, boolean> };
  ItemEditor: typeof ItemEditorDefault;
  errorCallback?: (message: string, e: Error) => void;
  PopupPanel: IPopupPanelProps;
  notificationDispatch: (message: string) => void;
  customParsers?: { [k: string]: CustomParser };
  previewHeadings?: IHeading[];
  asyncGetDocument: TAsyncGetDocument<ResultType>;
  asyncPutDocument: TAsyncPutDocument<ResultType>;
  asyncPostDocument: TAsyncPostDocument<ResultType>;
  asyncGetDocuments: TAsyncGetDocuments<ResultType>;
  asyncDeleteDocument: TAsyncDeleteDocument;
  asyncCopyDocument: TAsyncCopyDocument;
  componentMap: TComponentMap;
  closePopupOnItemSave?: boolean;
  sasProps?: Partial<ISearchAndSelectProps<ResultType>>;
}

/**
 * Generic catalogue wrapper for searching and editing documents from the api
 * @param id
 * @param itemName
 * @param collection
 * @param additionalFilters
 * @param customSort
 * @param resultsHeadings
 * @param editorHeadings
 * @param additionalSaveData
 * @param availableFilters
 * @param ItemEditor
 * @param errorCallback
 * @param PopupPanel
 * @param notificationDispatch
 * @param customParsers
 * @param previewHeadings
 * @param componentMap Map of field component for item editor
 * @param {async function} asyncGetDocument async function to get a document
 * - @argument { string } collection target database collection
 * - @argument { string } uid id of document
 * - @argument { Array[string] } schema fields to request from db
 * - @argument { bool } populate if true populate nested docs
 * @param {async function} asyncGetDocuments async function to search for multiple documents
 * - @argument { string } collection target database collection
 * - @argument { List[FilterObjectClass] } filters list of filters to apply to search
 * - @argument { List[string] } schema list of fields to request
 * - @argument { string } sortBy Field to sort by
 * - @argument { string } textsearch string to use for text(Keyword) searching
 * - @argument { bool } reverseSort if true reverse sorting
 * - @argument { bool } populate if true populate nested docs
 * @param {async function} asyncPutDocument async function to update a document
 * - @argument { string } collection target database collection
 * - @argument { string } uid id of document
 * - @argument { object } data data to upload
 * @param {async function} asyncPostDocument async function to add a new document
 * - @argument { string } collection target database collection
 * - @argument { string } uid id of document
 * - @argument { object } data data to upload
 * @param {async function} asyncDeleteDocument async function to delete a document
 * - @argument { string } collection target database collection
 * - @argument { string } uid id of document
 * @param {async function} asyncCopyDocument async function to copy a document
 * - @argument { string } fromCollection db collection to copy from
 * - @argument { string } fromUid db uid to copy
 * - @argument { string } toCollection db collection to copy to
 * - @argument { string } toUid new document id
 * - @argument { object } additionalData additional data to add to copied doc
 * @returns
 */
export const GenericCatalogue = <ResultType extends IDocument>({
  id,
  itemName,
  collection,
  additionalFilters,
  customSort,
  resultsHeadings,
  editorHeadings,
  additionalSaveData,
  availableFilters,
  ItemEditor,
  errorCallback,
  PopupPanel,
  notificationDispatch,
  customParsers,
  previewHeadings,
  asyncGetDocument,
  asyncGetDocuments,
  asyncPutDocument,
  asyncPostDocument,
  asyncDeleteDocument,
  asyncCopyDocument,
  componentMap,
  closePopupOnItemSave,
  sasProps,
}: IGenericCatalogueProps<ResultType>) => {
  const [showEditor, setShowEditor] = useState(false);
  const [selectedUid, setSelectedUid] = useState<Uid | null>(null);
  const [reloadDatakey, setReloadDataKey] = useState(0);
  const [handleDelete, setHandleDelete] = useState(false);

  /* Handle Delete */
  useEffect(() => {
    if (handleDelete && selectedUid) {
      setHandleDelete(false);
      asyncDeleteDocument(collection, selectedUid)
        .then(() => {
          setReloadDataKey((prev) => prev + 1);
        })
        .catch((e) => {
          if (errorCallback) errorCallback(`Error deleting ${itemName}`, e);
        });
    }
  }, [handleDelete, selectedUid, errorCallback, collection, itemName]);

  const handleSasSelect = useCallback((data?: ResultType | null) => {
    if (data) setSelectedUid(data.uid);
    else setSelectedUid(null);
  }, []);

  const duplicateCallback = useCallback(
    (_, [, , , toUid]) => {
      setReloadDataKey((prev) => prev + 1);
      notificationDispatch(`Successfully Copied ${itemName}`);
      setSelectedUid(toUid);
      setShowEditor(true);
    },
    [notificationDispatch, itemName]
  );

  /* Handle Duplicate */
  const { call: copyItem } = useAsyncRequest<void, any[]>({
    args: [],
    callFn: asyncCopyDocument as any,
    callOnInit: false,
    callback: duplicateCallback,
  });

  const handleDuplicate = (uid) => {
    const fromCollection = collection;
    const fromUid = uid;
    const toCollection = collection;
    const toUid = generateUid(collection);
    const additionalData = { label: toUid };
    copyItem([fromCollection, fromUid, toCollection, toUid, additionalData]);
  };

  /* Handle Search */

  const searchFn = useCallback(
    async (filters, sortBy, searchString) => {
      const sortByA = sortBy || 'label';
      const docs = await asyncGetDocuments(
        collection,
        filters,
        // TODO: Filter preview only headings out
        resultsHeadings.map((item) => item.uid),
        sortByA,
        searchString,
        false,
        'all'
      );
      if (customSort) return docs.sort(customSort);
      return docs;
    },
    [collection, resultsHeadings]
  );

  const onSubmitCallback = useCallback(
    (uid) => {
      notificationDispatch(`${itemName} saved`);
      setSelectedUid(uid);
      if (closePopupOnItemSave) setShowEditor(false);
      setReloadDataKey((prev) => prev + 1);
    },
    [itemName]
  );

  return (
    <>
      {showEditor && (
        <ItemEditor
          inputUid={selectedUid}
          isNew={!selectedUid && true}
          additionalData={additionalSaveData}
          params={editorHeadings}
          collection={collection}
          onSubmitCallback={onSubmitCallback}
          asyncGetDocument={asyncGetDocument}
          asyncPutDocument={asyncPutDocument}
          asyncPostDocument={asyncPostDocument}
          asyncDeleteDocument={asyncDeleteDocument}
          componentMap={componentMap}
          saveErrorCallback={errorCallback}
          // onCancel={() => setShowEditor(false)}
          id={`item_editor_${id}`}
        />
      )}
      <div className={`genericCatalogueFunc_Wrap sectionWrapper genericCatalogue_${id}`}>
        <section>
          <button
            type="button"
            className="button-two genericCatalogue_createNewBtn"
            onClick={() => {
              setSelectedUid(null);
              setShowEditor(true);
            }}
          >
            Create New {itemName}
          </button>
        </section>
        <section style={{ width: '100%' }}>
          <p>
            <b>Adjust the filters below to search for {itemName}s</b>
          </p>
          <SearchAndSelect
            key={reloadDatakey}
            searchFunction={searchFn}
            initialFilters={additionalFilters}
            availableFilters={availableFilters}
            handleSelect={handleSasSelect}
            headings={resultsHeadings}
            previewHeadings={previewHeadings || resultsHeadings}
            customParsers={customParsers}
            autoUpdate
            showRefreshBtn
            showSearchField
            loadOnInit={false}
            noEmptySearch
            allowSelectionPreview
            {...sasProps}
          />
          <div>
            <button
              type="button"
              className="button-two genericCatalogue_editSelectedBtn"
              onClick={() => setShowEditor(true)}
              disabled={selectedUid === null}
            >
              Edit Selected {itemName}
            </button>
            <button
              type="button"
              className="button-one genericCatalogue_duplicateSelectedBtn"
              onClick={() => handleDuplicate(selectedUid)}
              disabled={selectedUid === null}
            >
              Duplicate Selected {itemName}
            </button>
            <AreYouSureBtn
              onConfirmed={() => {
                setHandleDelete(true);
              }}
              message={`Delete ${itemName}`}
              disabled={selectedUid === null}
              btnText={<Emoji emoj="🗑️" label="Delete" />}
              PopupPanel={PopupPanel}
            />
          </div>
        </section>
      </div>
    </>
  );
};

GenericCatalogue.propTypes = {
  id: PropTypes.string.isRequired,
  itemName: PropTypes.string.isRequired,
  collection: PropTypes.string.isRequired,
  additionalFilters: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      field: PropTypes.string.isRequired,
      operator: PropTypes.string,
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    })
  ),
  customSort: PropTypes.func,
  customParsers: PropTypes.objectOf(PropTypes.func),
  resultsHeadings: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  editorHeadings: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  previewHeadings: PropTypes.arrayOf(PropTypes.shape({})),
  additionalSaveData: PropTypes.shape({}),
  availableFilters: PropTypes.objectOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
      required: PropTypes.bool,
      group: PropTypes.number,
    })
  ).isRequired,
  ItemEditor: PropTypes.elementType,
  PopupPanel: PropTypes.elementType,
  notificationDispatch: PropTypes.func,
  asyncGetDocuments: PropTypes.func.isRequired,
  asyncDeleteDocuments: PropTypes.func.isRequired,
  asyncGetDocument: PropTypes.func.isRequired,
  asyncPutDocument: PropTypes.func.isRequired,
  asyncPostDocument: PropTypes.func.isRequired,
  asyncDeleteDocument: PropTypes.func.isRequired,
  componentMap: PropTypes.objectOf(PropTypes.elementType),
  errorCallback: PropTypes.func,
  closePopupOnItemSave: PropTypes.bool,
};

GenericCatalogue.defaultProps = {
  additionalFilters: [],
  additionalSaveData: {},
  customSort: null,
  previewHeadings: [],
  customParsers: {},
  ItemEditor: ItemEditorDefault,
  PopupPanel: () => {},
  notificationDispatch: alert,
  componentMap: {},
  errorCallback: () => {},
  closePopupOnItemSave: false,
};
