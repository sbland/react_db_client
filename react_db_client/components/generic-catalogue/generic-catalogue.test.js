import '@samnbuk/react_db_client.helpers.enzyme-setup';
import React from 'react';
import { mount } from 'enzyme';
import { render, fireEvent, within, act } from '@testing-library/react';
import { MockReactC } from '@samnbuk/react_db_client.helpers.testing';
import { BasicGenericCatalogue } from './generic-catalogue.composition';

// jest.mock('@samnbuk/react_db_client.components.search-and-select', () =>
//   MockReactC('SearchAndSelect', ['SearchAndSelect'])
// );

const defaultProps = {};

describe('CertificateCatalogue', () => {
  it('Renders', () => {
    mount(<BasicGenericCatalogue />);
  });
  it('Matches Snapshot', () => {
    const component = mount(<BasicGenericCatalogue />);
    const tree = component.debug();
    expect(tree).toMatchSnapshot();
  });
  describe('Functional tests', () => {
    test('should render list of results', async () => {
      const component = render(<BasicGenericCatalogue />);
      const searchInput = component.getByLabelText('search');
      await act(async () => {
        fireEvent.change(searchInput, { target: { value: 'demo' } });
        await new Promise((resolve) => setTimeout(resolve));
      });
      const resultsList = component.getByRole('list');
      const { getAllByRole } = within(resultsList);
      const items = getAllByRole('listitem');
      expect(items.length).toBeGreaterThan(0);
      expect(items.length).toBe(5);
    });
  });
});
