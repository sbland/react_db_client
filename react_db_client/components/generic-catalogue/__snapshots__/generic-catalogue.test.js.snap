// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CertificateCatalogue Matches Snapshot 1`] = `
"<BasicGenericCatalogue>
  <GenericCatalogue id=\\"demo-id\\" itemName=\\"Demo Item\\" collection=\\"DEMO COLLECTION\\" additionalFilters={{...}} availableFilters={{...}} resultsHeadings={{...}} editorHeadings={{...}} additionalSaveData={{...}} ItemEditor={[Function: ItemEditor]} apiDeleteDocuments={[Function: apiDeleteDocuments]} PopupPanel={[Function: PopupPanel]} notificationDispatch={[Function (anonymous)]} asyncGetDocument={[Function: asyncGetDocument]} asyncGetDocuments={[Function: asyncGetDocuments]} asyncPutDocument={[Function: asyncPutDocument]} asyncPostDocument={[Function: asyncPostDocument]} asyncDeleteDocument={[Function: asyncDeleteDocument]} asyncDeleteDocuments={[Function: asyncDeleteDocuments]} componentMap={{...}} onError={[Function: onError]}>
    <div className=\\"GenericCatalogueFunc_Wrap sectionWrapper\\">
      <section>
        <button type=\\"button\\" className=\\"button-two\\" onClick={[Function: onClick]}>
          Create New 
          Demo Item
        </button>
      </section>
      <section style={{...}}>
        <p>
          <b>
            Adjust the filters below to search for 
            Demo Item
            s
          </b>
        </p>
        <SearchAndSelect searchFunction={[Function (anonymous)]} initialFilters={{...}} availableFilters={{...}} handleSelect={[Function: bound dispatchAction]} headings={{...}} previewHeadings={{...}} autoWidth={false} autoUpdate={true} showRefreshBtn={true} showSearchField={true} loadOnInit={false} noEmptySearch={true} allowSelectionPreview={true} selectionOverride={{...}} allowFilters={true} allowMultiple={false} returnFieldOnSelect=\\"uid\\" searchFieldTargetField={{...}} acceptSelectionBtnText=\\"Accept Selection\\" limitResultHeight={{...}} sortBy=\\"uid\\" reverseSort={false} reloadKey={0} liveUpdate={false} customParsers={{...}} labelField=\\"label\\" autoPreview={false}>
          <div className=\\"searchAndSelect sas_wrap sectionWrapper\\">
            <section className=\\"sas_filtersSection\\" style={{...}}>
              <button type=\\"button\\" className=\\"button-one\\" onClick={[Function: onClick]}>
                Show Preview
              </button>
              <FilterPanel filterData={{...}} clearFilters={[Function: handleClearFilters]} addFilter={[Function: addFilter]} deleteFilter={[Function: deleteFilter]} updateFilter={[Function: updateFilter]} fieldsData={{...}} showPanelOverride={false} floating={false} autoOpenPanel={false} customFilters={{...}} customFiltersComponents={{...}}>
                <div className=\\"filterManager\\">
                  <button type=\\"button\\" className=\\"button-one openFiltersButton\\" onClick={[Function: onClick]}>
                    Filters
                  </button>
                  <div className=\\"filterPanel_panel \\" style={{...}}>
                    <FiltersList filterData={{...}} deleteFilter={[Function: deleteFilter]} updateFilter={[Function: updateFilter]} fieldsData={{...}} customFilters={{...}} customFiltersComponents={{...}}>
                      <ul className=\\"filterPanel_filterList\\">
                        No filters
                      </ul>
                    </FiltersList>
                    <button type=\\"button\\" className=\\"button-one\\" onClick={[Function: onClick]}>
                      Clear Filters
                    </button>
                    <AddFilterButton fieldsData={{...}} returnNewFilter={[Function: returnNewFilter]} customFilters={{...}}>
                      <button type=\\"button\\" className=\\"button-two addFilterBtn\\" onClick={[Function: handleNewFilter]}>
                        Add Filter
                      </button>
                    </AddFilterButton>
                  </div>
                </div>
              </FilterPanel>
              <div style={{...}}>
                <label>
                  Search: 
                </label>
                <input className=\\"searchField\\" style={{...}} aria-label=\\"search\\" type=\\"text\\" placeholder=\\"search...\\" value=\\"\\" onChange={[Function: handleSearchFieldInput]} />
              </div>
              <button type=\\"button\\" className=\\"button-reset refreshBtn\\" onClick={[Function: onClick]} style={{...}}>
                <Emoji emoj=\\"üîÑ\\" label=\\"refresh\\" className=\\"\\">
                  <span className=\\"emoji \\" role=\\"img\\" aria-label=\\"refresh\\">
                    üîÑ
                  </span>
                </Emoji>
              </button>
            </section>
            <section className=\\"sas_resultsSection\\" style={{...}}>
              <div className=\\"sas_resultsList\\">
                <div style={{...}}>
                  <StyledSelectList listInput={{...}} headings={{...}} handleSelect={[Function: handleItemSelect]} currentSelection={{...}} limitHeight={{...}} selectionField=\\"uid\\" autoWidth={false} customParsers={{...}}>
                    <div className=\\"styledList\\" style={{...}}>
                      <div style={{...}} className=\\"styledList_headings\\">
                        <div className=\\"styledList_heading\\" style={{...}}>
                          UID
                        </div>
                        <div className=\\"styledList_heading\\" style={{...}}>
                          Name
                        </div>
                        <div className=\\"styledList_heading\\" style={{...}}>
                          Count
                        </div>
                        <div className=\\"styledList_heading\\" style={{...}}>
                          Eval
                        </div>
                      </div>
                      <div style={{...}} className=\\"styledList_items \\" role=\\"list\\" />
                    </div>
                  </StyledSelectList>
                </div>
                <div className=\\"sas_resultsList-empty\\">
                  No results found. Try adjusting the filters above.
                </div>
              </div>
            </section>
          </div>
        </SearchAndSelect>
        <div>
          <button type=\\"button\\" className=\\"button-two\\" onClick={[Function: onClick]} disabled={true}>
            Edit Selected 
            Demo Item
          </button>
          <AreYouSureBtn onConfirmed={[Function: onConfirmed]} message=\\"Delete Demo Item\\" disabled={true} btnText={{...}} PopupPanel={[Function: PopupPanel]} className=\\"button-one\\" notes=\\"\\" confirmMessage=\\"Confirm\\">
            <PopupPanel isOpen={false} handleClose={[Function: handleClose]} />
            <button disabled={true} onClick={[Function: onClick]} type=\\"button\\" className=\\"areYouSureBtn button-one\\">
              <Emoji emoj=\\"üóëÔ∏è\\" label=\\"Delete\\" className=\\"\\">
                <span className=\\"emoji \\" role=\\"img\\" aria-label=\\"Delete\\">
                  üóëÔ∏è
                </span>
              </Emoji>
            </button>
          </AreYouSureBtn>
        </div>
      </section>
    </div>
  </GenericCatalogue>
</BasicGenericCatalogue>"
`;
